src/deps/uws.zig

```diff
@@ -69,21 +73,455 @@ pub const InternalLoopData = extern struct {
+            const this = @as(*UpgradedDuplex, @ptrCast(@alignCast(self)));
+
+            if (this.wrapper != null) {
+                this.handlers.onEnd(this.handlers.ctx);
+            }
+        }
+    }
+
+    fn onWritable(
+        globalObject: *JSC.JSGlobalObject,
+        callframe: *JSC.CallFrame,
+    ) JSC.JSValue {
+        log("onWritable", .{});
+
+        _ = globalObject;
+        const function = callframe.callee();
+
+        if (JSC.getFunctionData(function)) |self| {
+            const this = @as(*UpgradedDuplex, @ptrCast(@alignCast(self)));
+            // flush pending data
+            if (this.wrapper) |*wrapper| {
+                _ = wrapper.flush();
+            }
+            // call onWritable (will flush on demand)
+            this.handlers.onWritable(this.handlers.ctx);
+        }
+
+        return JSC.JSValue.jsUndefined();
+    }
+
+    fn onCloseJS(
+        globalObject: *JSC.JSGlobalObject,
+        callframe: *JSC.CallFrame,
+    ) JSC.JSValue {
+        log("onCloseJS", .{});
+
+        _ = globalObject;
+        const function = callframe.callee();
+
+        if (JSC.getFunctionData(function)) |self| {
+            const this = @as(*UpgradedDuplex, @ptrCast(@alignCast(self)));
+            // flush pending data
+            if (this.wrapper) |*wrapper| {
+                _ = wrapper.shutdown(true);
+            }
+        }
+
+        return JSC.JSValue.jsUndefined();
+    }
+
+    pub fn onTimeout(this: *UpgradedDuplex) EventLoopTimer.Arm {
+        log("onTimeout", .{});
+
+        const has_been_cleared = this.event_loop_timer.state == .CANCELLED or this.vm.scriptExecutionStatus() != .running;
+
+        this.event_loop_timer.state = .FIRED;
+        this.event_loop_timer.heap = .{};
+
+        if (has_been_cleared) {
+            return .disarm;
+        }
+
+        this.handlers.onTimeout(this.handlers.ctx);
+
+        return .disarm;
+    }
+
+    pub fn from(
+        globalThis: *JSC.JSGlobalObject,
+        origin: JSC.JSValue,
+        handlers: UpgradedDuplex.Handlers,
+    ) UpgradedDuplex {
+        return UpgradedDuplex{
+            .vm = globalThis.bunVM(),
+            .origin = JSC.Strong.create(origin, globalThis),
+            .wrapper = null,
+            .handlers = handlers,
+        };
+    }
+
+    pub fn startTLS(this: *UpgradedDuplex, ssl_options: JSC.API.ServerConfig.SSLConfig, is_client: bool) !void {
+        this.wrapper = try WrapperType.init(ssl_options, is_client, .{
+            .ctx = this,
+            .onOpen = UpgradedDuplex.onOpen,
+            .onHandshake = UpgradedDuplex.onHandshake,
+            .onData = UpgradedDuplex.onData,
+            .onClose = UpgradedDuplex.onClose,
+            .write = UpgradedDuplex.internalWrite,
+        });
+        if (this.origin.get()) |duplex| {
+            const globalThis = this.origin.globalThis.?;
+
+            const addEventListener = try duplex.getFunction(globalThis, "on") orelse return error.InvalidDuplex;
```

Comment:
Can we do this step in JavaScript instead? Calling functions like this and using runCallback here will not handle exceptions properly

Reply:
done